eth2-ssv/beacon/prysmgrpc/prysmgrpc.go

```diff
@@ -0,0 +1,260 @@
+	if duty == nil {
+		return nil, nil
+	}
+
+	var roles []beacon.Role
+	if len(duty.ProposerSlots) > 0 {
+		for _, proposerSlot := range duty.ProposerSlots {
+			if proposerSlot != 0 && proposerSlot == slot {
+				roles = append(roles, beacon.RoleProposer)
+				break
+			}
+		}
+	}
+	if duty.AttesterSlot == slot {
+		roles = append(roles, beacon.RoleAttester)
+
+		aggregator, err := b.isAggregator(ctx, slot, len(duty.Committee), shareKey)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not check if a validator is an aggregator")
+		}
+
+		if aggregator {
+			roles = append(roles, beacon.RoleAggregator)
+		}
+	}
+
+	if len(roles) == 0 {
+		roles = append(roles, beacon.RoleUnknown)
+	}
+
+	return roles, nil
+}
+
+// receiveBlocks receives blocks from the beacon node. Upon receiving a block, the service
+// broadcasts it to a feed for other usages to subscribe to.
+func (b *prysmGRPC) receiveBlocks(ctx context.Context) error {
+	stream, err := b.beaconClient.StreamBlocks(ctx, &ethpb.StreamBlocksRequest{VerifiedOnly: true})
+	if err != nil {
+		return errors.Wrap(err, "failed to open stream blocks")
+	}
+
+	for {
+		if ctx.Err() == context.Canceled {
+			return errors.New("context canceled - shutting down blocks receiver")
+		}
+		res, err := stream.Recv()
+		if err != nil {
+			if e, ok := status.FromError(err); ok {
+				switch e.Code() {
+				case codes.Canceled, codes.Internal, codes.Unavailable:
+					b.logger.Info("Trying to restart connection.", zap.Any("rpc status", e.Code()))
+					err = b.restartBeaconConnection(ctx)
+					if err != nil {
+						return errors.Wrap(err, "Could not restart beacon connection")
+					}
+					stream, err = b.beaconClient.StreamBlocks(ctx, &ethpb.StreamBlocksRequest{VerifiedOnly: true} /* Prefers unverified block to catch slashing */)
+					if err != nil {
+						return errors.Wrap(err, "Could not restart block stream")
+					}
+					b.logger.Info("Block stream restarted...")
+					continue // force loop again to recv
+				default:
+					return errors.WithMessagef(err, "Could not receive block from beacon node. rpc status - %v", e.Code())
+				}
+			}
+			return errors.Wrap(err, "failed to receive blocks from the node")
+		}
+
+		if res == nil || res.Block == nil {
+			b.logger.Debug("empty block resp")
+			continue
+		}
+
+		if res.Block.Slot > b.highestValidSlot {
+			b.highestValidSlot = res.Block.Slot
+		}
+
+		//b.logger.Debug("Received block from beacon node", TODO need to set trace level
+		//	zap.Any("slot", res.Block.Slot),
+		//	zap.Any("proposer_index", res.Block.ProposerIndex))
+		b.blockFeed.Send(res)
+	}
+}
+
+// domainData returns domain data for the given epoch and domain
+func (b *prysmGRPC) domainData(ctx context.Context, slot uint64, domain []byte) (*ethpb.DomainResponse, error) {
+	req := &ethpb.DomainRequest{
+		Epoch:  b.network.EstimatedEpochAtSlot(slot),
+		Domain: domain,
+	}
+
+	// TODO: Try to get data from cache
```

Comment:
Recommend TODOs with an assigned issue number to not forget and accumulate technical debt, i.e. TODO(#332): Use a cache